### DISCUSSION TITLE
Preserve original filename

### DISCUSSION THREAD

**Discussion by morozig** [thundaga/SD-webui-txt2img-script#22] [General]
Hi! Thanks for the script! I'm using it to do batch hi-res fix from files in a directory. But there is one feature I'm missing, I want to preserve original file names(or at least their seed) because I've got a lot of them in a directory.
For example: original files:
```
00009-84108608.png
00026-1617140356.png
00047-2981297943.png
```
Processed files:
```
0000-00009-84108608.png
0001-00026-1617140356.png
0002-00047-2981297943.png
```
I've tried to do it myself:
```
if tab_index == 1 and output_dir != '':
                for n, processed_image in enumerate(proc.images):
                    images.save_image(image=processed_image, path=output_dir, basename=proc.all_seeds[n], existing_info=processed_image.info)
```
But it inserted only seeds and lost counter for some reason. I'm not very good at python. Can you please help me with this functionality?

**Comment by OPCOXE** [comment]
I had the issue a while ago and couldn't fix it so I decided to create a rename script because it was easier for me.
I might check into it when I have time but don't give it an hope.

Here is what I'm currently using : 
 - A powershell script to rename de files :
```POWERSHELL
[CmdletBinding()]
param (
    [Parameter(Mandatory)]
    [String]
    $FolderToRename,
    [Parameter(Mandatory)]
    [String]
    $FolderRenameFrom
)

Set-Location $PSScriptRoot # New version

$ToRename = Get-ChildItem "$FolderToRename\*.png"
$RenameFrom = Get-ChildItem "$FolderRenameFrom\*.png"

for ($i = 0; $i -lt $ToRename.Count; $i++)
{
    Rename-Item "$FolderToRename\$($ToRename[$i].Name)" -NewName $RenameFrom[$i].Name
}
```
 - And since the script can't be executed with default policy I created a batch file that simply as the command to start the script with bypass policy :
 ```BATCH
 powershell -ExecutionPolicy Bypass -File .\Rename.ps1 .\ToRename\ .\RenameFrom\
```
Where ToRename is the folder with the new files (like 0000.png) and RenameFrom the folder with the old one (like 00000-21425415)

**Note : Do not use code from anyone without knowing what you do**

Edit : I think you could do an issue. I don't know if thundaga intend to let it the way it is or not

**Reply by morozig** [reply]
That's a great workaround, I've used your powershell script! I added a little sanity check to compare file counts and I've combined both filenames as in my original post. Here's final code:

`Rename.ps1`

```powershell
[CmdletBinding()]
param (
    [Parameter(Mandatory)]
    [String]
    $FolderToRename,
    [Parameter(Mandatory)]
    [String]
    $FolderRenameFrom
)

Set-Location $PSScriptRoot # New version

$ToRename = Get-ChildItem "$FolderToRename\*.png"
$RenameFrom = Get-ChildItem "$FolderRenameFrom\*.png"

if ($ToRename.Count -ne $RenameFrom.Count) {
  Write-Host "Files count mismatch!"
  exit 1
}

for ($i = 0; $i -lt $ToRename.Count; $i++)
{
    Rename-Item "$FolderToRename\$($ToRename[$i].Name)" -NewName "$($ToRename[$i].Basename)-$($RenameFrom[$i].Name)"
}

Write-Host "Done!"

```

**Comment by morozig** [comment]
Thanks. I still feel like this can be done in the original script with couple lines of code, I just don't know where to put them. I didn't start an issue because I'm not sure everyone would want that behaviour and to make this feature opt-in would require more work, that's why I just wanted to know those couple lines of code and where to put them.

**Reply by OPCOXE** [reply]
I personnaly think that having the name of the original image for the output image would be better than what it is right now.
The line you modified is what should be modified for sure, what should be investigated is how to get back the original image name (if possible).
I think I'll check that this week-end I have a few ideas I want to test

**Reply by OPCOXE** [reply]
My ideas didn't work but I found something that I can use to make it work.
I'm testing it

**Comment by OPCOXE** [comment]
Ok I found the solution :
First I remade the build_file_list function to put the filename into the infotext of the image
```PYTHON
def build_file_list(file, tab_index: int, file_list: list[dict]) -> list[dict]:

    file = file.name if tab_index == 0 else file
    file_ext = pathlib.Path(file).suffix
    filename = pathlib.Path(file).name.split(".")[0]

    if file_ext == ".txt":
        text = open(file, "r", encoding="utf-8").read()
    elif run_pnginfo(Image.open(file))[1] != None:
        text = run_pnginfo(Image.open(file))[1]
    
    if text != None and text != "":
        parsed_text = parse_generation_parameters(text)
        parsed_text["filename"] = filename
        file_list.append(parsed_text)

    return file_list
```
Then I use the value on the save_image function call
This generate the image with the exact same name
```PYTHON
images.save_image(image=processed_image, path=output_dir, basename='', forced_filename=parsed_text["filename"], existing_info=processed_image.info)
```
and this generate the image with "-0000" at the end (usefull when multiple input with same name like png, jpg and/or txt inputs with the same name)
```PYTHON
images.save_image(image=processed_image, path=output_dir, basename=parsed_text["filename"], existing_info=processed_image.info)
```
EDIT : This won't work when using the "Batch Process"

**Reply by morozig** [reply]
Thanks! I guess we have two solutions now. One is general, one for windows. Thanks for you time!

**Reply by OPCOXE** [reply]
I made a fork of the repo and added a dropdown for the filename format here : https://github.com/OPCOXE/SD-webui-txt2img-script
I'll send a pull request.
With this all the options are available
